lib/handler/status/durations.c

```diff
@@ -0,0 +1,231 @@
+    stats->connect_time = gkc_summary_alloc(GK_EPSILON);
+    stats->header_time = gkc_summary_alloc(GK_EPSILON);
+    stats->body_time = gkc_summary_alloc(GK_EPSILON);
+    stats->request_total_time = gkc_summary_alloc(GK_EPSILON);
+    stats->process_time = gkc_summary_alloc(GK_EPSILON);
+    stats->response_time = gkc_summary_alloc(GK_EPSILON);
+    stats->duration = gkc_summary_alloc(GK_EPSILON);
+}
+
+static void *durations_status_init(void)
+{
+    struct st_duration_agg_stats_t *agg_stats;
+
+    agg_stats = h2o_mem_alloc(sizeof(*agg_stats));
+
+    duration_stats_init(&agg_stats->stats);
+    pthread_mutex_init(&agg_stats->mutex, NULL);
+
+    return agg_stats;
+}
+
+static void duration_stats_free(struct st_duration_stats_t *stats)
+{
+    gkc_summary_free(stats->connect_time);
+    gkc_summary_free(stats->header_time);
+    gkc_summary_free(stats->body_time);
+    gkc_summary_free(stats->request_total_time);
+    gkc_summary_free(stats->process_time);
+    gkc_summary_free(stats->response_time);
+    gkc_summary_free(stats->duration);
+}
+
+static h2o_iovec_t durations_status_final(void *priv, h2o_globalconf_t *gconf, h2o_req_t *req)
+{
+    struct st_duration_agg_stats_t *agg_stats = priv;
+    h2o_iovec_t ret;
+
+#define BUFSIZE 16384
+    ret.base = h2o_mem_alloc_pool(&req->pool, BUFSIZE);
+    ret.len = snprintf(ret.base, BUFSIZE, ",\n"
+#define DURATION_FMT(x) \
+            " \"" #x "-0\": %" PRIu64 ",\n" \
+            " \"" #x "-25\": %" PRIu64 ",\n" \
+            " \"" #x "-50\": %" PRIu64 ",\n" \
+            " \"" #x "-75\": %" PRIu64 ",\n" \
+            " \"" #x "-99\": %" PRIu64 "\n"
+            DURATION_FMT(connect-time) ","
+            DURATION_FMT(header-time) ","
+            DURATION_FMT(body-time) ","
+            DURATION_FMT(request-total-time) ","
+            DURATION_FMT(process-time) ","
+            DURATION_FMT(response-time) ","
+            DURATION_FMT(duration),
+#undef DURATION_FMT
+#define DURATION_VALS(x) gkc_query(agg_stats->stats.x, 0), gkc_query(agg_stats->stats.x, 0.25), \
+                        gkc_query(agg_stats->stats.x, 0.5), gkc_query(agg_stats->stats.x, 0.75), \
+                        gkc_query(agg_stats->stats.x, 0.99)
+
+                        DURATION_VALS(connect_time), DURATION_VALS(header_time),
+                        DURATION_VALS(body_time), DURATION_VALS(request_total_time),
+                        DURATION_VALS(process_time), DURATION_VALS(response_time),
+                        DURATION_VALS(duration));
+#undef DURATION_VALS
+#undef BUFSIZE
+
+    duration_stats_free(&agg_stats->stats);
+    pthread_mutex_destroy(&agg_stats->mutex);
+
+    free(agg_stats);
+    return ret;
+}
+
+static int64_t diff_timeval(struct timeval *from, struct timeval *until)
+{
+    int32_t delta_sec = (int32_t)until->tv_sec - (int32_t)from->tv_sec;
+    int32_t delta_usec = (int32_t)until->tv_usec - (int32_t)from->tv_usec;
+    int64_t delta = (int64_t)(delta_sec * 1000*1000L) + delta_usec;
```

Comment:
You might want to convert either of `delta_sec` or `1000*1000L` to `int64_t` before multiplication, so that the operation would be done in 64-bit on ILP32.